control-flow/index.md

```diff
@@ -1,6 +1,368 @@
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+randomDelayedResponse(1, text => console.log(text)); // outputs 1
+randomDelayedResponse(2, text => console.log(text)); // outputs 2
+randomDelayedResponse(3, text => console.log(text)); // outputs 3
+randomDelayedResponse(4, text => console.log(text)); // outputs 4
+// Who will win?
+```
+
+We can run the above code multiple times, and the result is unpredictable. It is _asynchronous_, in that the results do not arrive in order. To create a predictable, _synchronous_ flow using callbacks we'd need to nest them.
+
+```
+function randomDelayedResponse(text, callback) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+randomDelayedResponse(1, text => {
+  console.log(text);
+  randomDelayedResponse(2, text => {
+    console.log(text);
+    randomDelayedResponse(3, text => {
+      console.log(text);
+      randomDelayedResponse(4, text => {
+        console.log(text);
+      });
+    });
+  });
+}); // outputs "1 2 3 4"
+```
+
+Structuring these callbacks works. However this approach can create code becomes difficult to understand and maintain. Let's look at another way.
+
+## Promises
+
+To avoid this callback hell, we can use a different structure called _promises_. A promise is a function that returns a _resolved_ response which we can chain using `then`, or a _rejected_ response which we can `catch`. It follows this pattern:
+
+```
+runOutPromise() // returns a response
+  .then(response => {
+    return foo; // this could optionally be another promise
+  })
+  .then(foo => {
+    // they can be chained
+  })
+  .catch(error => {
+    // catch "reject" output here
+  });
+```
+
+We can rewrite our runner example using promises like so:
+
+```
+function randomDelayedResponse(text, callback) {
+  return new Promise((resolve, reject) => {
+    const timeOut = Math.floor(Math.random() * 10) + 1;
+    setTimeout(() => {
+      resolve(text); // Replacing the callback with "resolve"
+    }, timeOut * 10);
+  });
+}
+
+randomDelayedResponse(1)
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(2);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(3);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(4);
+  })
+  .then(res => {
+    console.log(res);
+  })
+  .catch(error => {
+    // handle error
+  })
+// outputs "1 2 3 4"
+```
+
+Running the above code should output the correct order, albeit with random delays between each number.
+
+We aren't handling errors in this example, but if our function was to encounter an error we would `reject('some error')` and this would then be picked up by the `catch`.
```

Comment:
Maybe demo the reject part of the promise?

```
function randomDelayedResponse(text, callback) {
  return new Promise((resolve, reject) => {
    const timeOut = Math.floor(Math.random() * 10) + 1;

     // randomly reject if we hit a number divisible by 42
     if (timeOut % 42 === 0) return reject();

    setTimeout(() => {
      resolve(text); // Replacing the callback with "resolve"
    }, timeOut * 10);
  });
}
```

Reply:
Cool, maybe something like this:

```
function randomDelayedResponse(text, callback) {
  return new Promise((resolve, reject) => {
    const timeOut = Math.floor(Math.random() * 100) + 1;
    setTimeout(() => {
      if (text) {
        resolve(text); // Replacing the callback with "resolve"
      }
      reject('No text provided!'); // Reject when there is an error
    }, timeOut);
  });
}
```

Reply:
Nice - much better example!

But that's going to always reject - do you need to do:
```
      if (!text) {
          reject('No text provided!'); // Reject when there is no text provided
      }
      resolve(text); // Replacing the callback with "resolve"
```

Reply:
I'm not sure it would - I mean the above is probably a better pattern but the `response` should be like a `return` and break before reaching the `reject`

Reply:
I'll update it as suggested as I do prefer to return on errors first as a matter of principle :D